design/ShaderLanguageFAQ.md

```diff
@@ -0,0 +1,157 @@
+
+
+## How do you pronounce “WGSL”?
+
+The W3C WebGPU Community Group does not prescribe a particular
+pronunciation of “WGSL”.
+
+You can say it letter-by-letter, like “W G S L”.
+
+You can say it as a single word, like “wiggzle”, “wiggles”, or “wigsel”.
+
+
+## What does a WGSL shader look like?
+
+WGSL has not yet been specified.
+However, early prototyping work indicates the language can have a look
+and feel as represented in the following examples:
+
+A simple fragment shader:
+
+
+```
+[[location 0]] var<out> gl_FragColor : vec4<f32>;
+
+fn main() -> void {
+    gl_FragColor = vec4<f32>(0.4, 0.4, 0.8, 1.0);
+    return;
+}
+entry_point fragment = main;
+```
+
+
+A code fragment showing if/else:
+
+
+```
+  if (value > 5) {
+    wvar = 14;
+  } elseif (value > 2) {
+    wvar = 12;
+  } else {
+    wvar = 0;
+  }
+```
+
+
+A code fragment showing a switch:
+
+
+```
+switch (a) {
+  case 1: {
+    # The break is implicit
+  }
+  case 2: {
+     fallthrough;
+  }
+  default: {
+  }
+}
+```
+
+
+A code fragment showing a loop:
+
+
+```
+var i : i32 = 0;
+var b : f32 = 2.4;
+loop {
+  break if (i == 4);
+
+  b = b * 3.2;
+  continuing {
+     i = i + 1;
+  }
+}
+```
+
+
+
+## Where can I find the spec?
+
+The WGSL spec will be developed on the WebGPU Community Group area on GitHub.
+See https://github.com/gpuweb/gpuweb/tree/master/wgsl
+
+
+## What browsers support WGSL?
+
+WGSL is the shader language for WebGPU.
+Any browser implementing WebGPU will be required to accept shaders in WGSL.
+Specifically, the WebGPU conformance test suite will use shaders written
+in WGSL.
+
+Browser support for WGSL is expected to be developed over time, along
+with the development of WGSL itself.
+
+
+## Will browsers accept other shader languages for WebGPU?
+
+Accepting a shader language other than WGSL is not part of WebGPU.
+Doing so harms interoperability, as a shader in that other language is
+not portable across browsers.
+
+
+## Where can I find WGSL tooling?
+
+The WebGPU Community Group does not deliver tooling for WGSL.
+However, WebGPU implementers are expected to develop tooling
+for handling WGSL, including parsing, validation, and compilation.
+
+At this time, we expect at least three essentially independent
+implementations of WGSL ingestion as part of browser support for WebGPU.
+Independent implementation serves as an important check on the quality
+of the WGSL and WebGPU specifications.
+
+
+## What about my existing shaders, compilers, and other tooling?
+
+This is also expressed as “the last thing I need is another shader language”.
+(See https://xkcd.com/927/)
+
+For developers invested in the SPIR-V ecosystem for Vulkan and OpenGL,
+the key will be convertibility between WGLSL and SPIR-V.
+For shaders that only use WebGPU features, WebGPU Community Group
+members expect to develop tooling to faithfully convert between WGSL
+and “Shader” SPIR-V.
+By “faithful”, we mean that all salient properties of a shader should
```

Comment:
Doesn't every compiler (intend to) faithfully preserve behavior for the subset of the source language it intends to support? This paragraph doesn't seem to mean anything, and I would delete it.

Reply:
This is attempting to say that a round trip of WGSL -> SPIR-V ->WGSL won't be byte for byte identical.

Reply:
The round tripping of WGSL -> MSL -> WGSL also won't be byte for byte identical. Also, the round tripping of WGSL -> HLSL -> WGLSL also won't be byte for byte identical.

Reply:
I mean something less than byte-for-byte identicality, but more than a typical round trip between shader languages.
I hinted at it with the salience of factors affecting "performance".  Details are TBD. But for example, the achievable parallelism should be preserved through the round trip I care about.  E.g. effective occupancy of a warp on NV hardware or wavefront on AMD hardware.  It's likely that a  good proxy for that is private register count, and the hypothesis is that can be faithfully counted by staring at WGSL source.  The goal is to avoid having a translation to/from WGSL induce falling off a performance cliff.  In my view that's critical to the success of WebGPU.

That's why the paragraph exists.